### PR TITLE
ClaimTransaction serialization

### PR DESCRIPTION
**Pull request by shargon**
- Refactor BinarySerialization
- Transaction and ClaimTransaction binary serialization
- More TX
- TransactionContext for computed values like SystemFee, networkFee or References

### CODE REVIEW

File: src/NeoSharp.BinarySerialization/BinaryDeserializer.cs [hunk 1]
```diff
@@ -39,7 +38,7 @@ public static void RegisterTypes(params Assembly[] asms)
         /// <param name="data">Data</param>
         /// <param name="settings">Settings</param>
         /// <returns>Return byte array</returns>
-        public T Deserialize<T>(byte[] data, BinarySerializerSettings settings = null) where T : new()
+        public T Deserialize<T>(byte[] data, BinarySerializerSettings settings = null)
```
Comment (osmirnov) [inline src/NeoSharp.BinarySerialization/BinaryDeserializer.cs]: Keep in my mind, Activator.CreateInstance is way more slower then new T()

File: src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs [hunk 2]
```diff
@@ -43,43 +41,52 @@ internal static BinarySerializerCache InternalRegisterTypes(Type type)
...
 
+        #endregion
+
         /// <summary>
         /// Type
         /// </summary>
         public readonly Type Type;
         /// <summary>
-        /// Count
+        /// IsEmpty
```
Comment (osmirnov) [inline src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs]: IsEmpty => _entries.Any() ?

File: src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs [hunk 3]
```diff
@@ -127,14 +142,19 @@ public BinarySerializerCache(Type type)
...
+            }
+
             var ret = 0;
-            var haveFilter = settings != null && settings.Filter != null; 
+            var haveFilter = settings != null && settings.Filter != null;
 
             foreach (BinarySerializerCacheEntry e in _entries)
             {
-                if (haveFilter && !settings.Filter.Invoke(e.Context)) continue;
+                if (haveFilter && !settings.Filter.Invoke(e.Context.Order)) continue;
```
Comment (osmirnov) [inline src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs]: What if the order will change?

File: src/NeoSharp.BinarySerialization/Extensions/BinaryReaderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        throw new FormatException();
+                    var count = groupSize - padding;
+                    if (count > 0)
+                        ms.Write(group, 0, count);
+                } while (padding == 0);
+                return ms.ToArray();
+            }
+        }
+
+        public static string ReadFixedString(this BinaryReader reader, int length)
```
Comment (osmirnov) [inline src/NeoSharp.BinarySerialization/Extensions/BinaryReaderExtensions.cs]: BinaryReader is capable to read fixed string if it was serialized before with length + payload. Why do we need this method?

File: src/NeoSharp.BinarySerialization/Extensions/BinaryWriterExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            for (var i = 0; i < padding; i++)
+                writer.Write((byte)0);
+            writer.Write((byte)padding);
+        }
+
+        public static void WriteFixedString(this BinaryWriter writer, string value, int length)
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+            if (value.Length > length)
```
Comment (osmirnov) [inline src/NeoSharp.BinarySerialization/Extensions/BinaryWriterExtensions.cs]: why can't we always count on value.length?

File: src/NeoSharp.BinarySerialization/BinaryDeserializer.cs (hunk 1 above)
Comment (shargon) [reply src/NeoSharp.BinarySerialization/BinaryDeserializer.cs]: The problem is the arrays. I need to remove this if i want to allow Desearialize<Transaction[]> or similar

File: src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs (hunk 2 above)
Comment (shargon) [reply src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs]: _entries are empty too when the cache use directly a `IBinaryCustomSerializer`

https://github.com/CityOfZion/neo-sharp/pull/125/files#diff-1e2f491d84304a3e8dae375555162e48R88

File: src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs (hunk 3 above)
Comment (shargon) [reply src/NeoSharp.BinarySerialization/Cache/BinarySerializerCache.cs]: Is a readonly field, should not change. It is used for allow to exclude some data, like signature or hash (for compute their hash or signature)

File: src/NeoSharp.BinarySerialization/Extensions/BinaryWriterExtensions.cs (hunk 5 above)
Comment (shargon) [reply src/NeoSharp.BinarySerialization/Extensions/BinaryWriterExtensions.cs]: This method is moved, i think that length is maxLength. I think that is not necessary, do you want to remove it?

**Comment by shargon** [comment]
Works for me ...

**Comment by osmirnov** [comment]
@aboimpinto could you take a look for coverall ssl connection?

**Comment by aboimpinto** [comment]
yes .. next task will be this ..

**Comment by shargon** [comment]
Can we merge this?

**Comment by osmirnov** [comment]
Ok I will merge this. Probably, it is some random error.

**Pull request merged**